packages/dd-trace/src/plugins/util/web.js

```diff
@@ -221,107 +247,85 @@ const web = {
 
-function finishMiddleware (context) {
-  if (context.finished) return
+    addRequestTags(context)
+    addResponseTags(context)
 
-  let span
+    context.config.hooks.request(context.span, req, res)
+    addResourceTag(context)
 
-  while ((span = context.middleware.pop())) {
-    span.finish()
-  }
-}
+    context.span.finish()
+    context.finished = true
+  },
+  wrapWriteHead (context) {
+    const { req, res } = context
+    const writeHead = res.writeHead
 
-function wrapEnd (context) {
-  const scope = context.tracer.scope()
-  const req = context.req
-  const res = context.res
-  const end = res.end
+    return function (statusCode, statusMessage, headers) {
+      headers = typeof statusMessage === 'string' ? headers : statusMessage
+      headers = Object.assign(res.getHeaders(), headers)
 
-  res.writeHead = wrapWriteHead(context)
+      if (req.method.toLowerCase() === 'options' && isOriginAllowed(req, headers)) {
+        addAllowHeaders(req, res, headers)
+      }
 
-  ends.set(res, function () {
-    for (const beforeEnd of context.beforeEnd) {
-      beforeEnd()
+      return writeHead.apply(this, arguments)
     }
+  },
+  wrapEnd (context) {
+    const scope = context.tracer.scope()
+    const req = context.req
+    const res = context.res
+    const end = res.end
 
-    finishMiddleware(context)
-
-    if (incomingHttpRequestEnd.hasSubscribers) incomingHttpRequestEnd.publish({ req, res })
-
-    const returnValue = end.apply(res, arguments)
+    res.writeHead = web.wrapWriteHead(context)
 
-    finish(context)
+    ends.set(res, function () {
+      for (const beforeEnd of context.beforeEnd) {
+        beforeEnd()
+      }
 
-    return returnValue
-  })
+      web.finishMiddleware(context)
 
-  Object.defineProperty(res, 'end', {
-    configurable: true,
-    get () {
-      return ends.get(this)
-    },
-    set (value) {
-      ends.set(this, scope.bind(value, context.span))
-    }
-  })
-}
+      if (incomingHttpRequestEnd.hasSubscribers) incomingHttpRequestEnd.publish({ req, res })
 
-function wrapWriteHead (context) {
-  const { req, res } = context
-  const writeHead = res.writeHead
+      const returnValue = end.apply(res, arguments)
 
-  return function (statusCode, statusMessage, headers) {
-    headers = typeof statusMessage === 'string' ? headers : statusMessage
-    headers = Object.assign(res.getHeaders(), headers)
+      web.finishSpan(context)
 
-    if (req.method.toLowerCase() === 'options' && isOriginAllowed(req, headers)) {
-      addAllowHeaders(req, res, headers)
-    }
+      return returnValue
+    })
 
-    return writeHead.apply(this, arguments)
+    Object.defineProperty(res, 'end', {
+      configurable: true,
+      get () {
+        return ends.get(this)
+      },
+      set (value) {
+        ends.set(this, scope.bind(value, context.span))
+      }
+    })
   }
 }
 
-function addAllowHeaders (req, res, headers) {
+function addAllowHeaders (req, headers) {
+  const res = req._datadog.res
```

Comment:
This property no longer exists. Are we missing a test? Why does it still work without it.